### ISSUE TITLE
230102 label propagation

### ISSUE THREAD

**Issue by gnkwon95** [gnkwon95/weeklyPaper#2]
### 논문 

Xiaojin Zhu and Zoubin Ghahramani. Learning from labeled and unlabeled data with label propagation
http://mlg.eng.cam.ac.uk/zoubin/papers/CMU-CALD-02-107.pdf

참고 미디엄: https://towardsdatascience.com/label-propagation-demystified-cd5390f27472

### 요약 설명

<img width="688" alt="image" src="https://user-images.githubusercontent.com/42033758/210202327-ae0db37f-4723-41ef-b1dd-90090f4f386a.png">

그래프에서 어느 목적지 노드들이 있고, 타겟 노드가 있을 때 edge들 (A -> B로 넘어갈 가능성)을 확률로 표기
해당 행렬 무한히 곱한 후
목적지 노드 (identity matrix)와 곱하면 특정 노드가 어느 목적지와 더 가까운지 값이 나옴

### 다른 설명

레이블 없는건 0으로 시작, 레이블된걸 통해 값이 유추되는 방식
a라는 웨이트값으로 업데이트된 레이블과 초기 레이블 (학습량)이 조절됨

### DGL

https://docs.dgl.ai/en/0.8.x/generated/dgl.nn.pytorch.utils.LabelPropagation.html
intent walk에선 조금 다르게 사용함

목표 지점의 값을 0으로 놓고 멀리 있는 레이블과 가장 가까운걸 찾는 (0인 문서의 소속을 찾는) 논문에 비해
본인을 1로 놓고 나머지를 0으로 놓고 본인제외 가장 값이 높은 노드를 찾는 방식
근데 dgl로도 구현은 가능할듯 

### 느낀점

생각보다 초기 논문이라 10분이면 읽고, 로직은 매우 간단함
다만 그래프로 직접 구현하려면... 또 다른 내용
dgl에서 라이브러리 사용 시 바로 확인할 수 있지만
그래프라는게 간단하게 쓱쓱 돌려보고 학습되는 과정을 체감하기가 어려운 것 같음

**Comment by gnkwon95** [comment]
## DGL 사용한 label propagation

질의, 세션아이디, 키워드에 대해 (gdid, weight)값을 생성한다
ㄴ 쉽게하기 위해 weight는 전부 1로 통일한다
ㄴ 단, 질의/세션아이디/키워드는 노이즈를 제거하고 많이 사용되는 것들만 사용한다

질의/세션아이디/키워드를 같은 칼럼명으로 넣어버린다 (entity, gdid, weight 테이블 생성)
dgl 그래프로 변환한다
ㄴ pinsage에 있는 pandasGraphMaker, graph_maker라는 두개 함수로 entity - gdid 간의 그래프를 만들고, edge 설정 가능하다
ㄴ 노드에 gdid 노드인지, entity 노드인지 레이블되어있다

개별 doc에 대해 subgraph 가져와서 가장 가까운 레이블값 가져오기
ㄴ 본인 gdid 노드는 1, 나머지는 0으로 표기, entity 노드는 전부 0으로 표기
ㄴ 모든 A - edge -> B 에 대해 
ㄴㄴ 본인값과 weight를 곱해서, m으로 전달. m을 합해서 h값 업데이트
ㄴㄴ 본인 gdid만 1, 나머진 모두 0에서 weight 따라 퍼져나가는 방식

일반 label propagation은 본인이 0이고, 다른 레이블에서 조금씩 퍼져와서 최종값을 가져오는 방식이지만
이 추천 방식은 본인이 1이고 나머지 0으로 시작해서 가장 높은 두번째 값을 찾는 방식

최종적으로 subgraph에서 본인은 빼고 gdid - 값을 튜플로 묶어서 반환 (gdid node만 한정)
